### PR TITLE
Flatter TTL data without levelDB

### PR DESCRIPTION
**Pull request by adiabat**
This PR removes levelDB usage for the bridge node when dealing with TTL data, and instead stores the TTL data in 2 flat files, txidFile and txidOffsetFile.  

It seems to work now with stop & restart, and with a testnet3 sync it works pretty quickly, and results in 440MB of extra data, which seems very reasonable.  There are ways to decrease this further (txidOffsetFile is mostly zeros, and compresses from 14MB to 4MB.  txidFile doesn't compress, but there's still a lot of structure there.)
Anyway 440MB seems like no big deal so probably not worth working on reducing that.

I've commented out the memTTL options in this PR, as there's a bit of work to get the map-based TTL working without depending on levelDB.  Depending on the speed, it might not be worth it to keep in the map-based code as the flat file seems pretty fast, at least on the SSD I'm using.

Note that levelDB is still in the bridge node code to read the bitcoin blocks metadata.  It might be fun to eliminate levelDB entirely, but it's really only used in the first few seconds to figure out how to read the bitcoin rev files.

### CODE REVIEW

**Comment by kcalvinalvin** [comment]
Regtest ci passes for me on a local build (after fixing some minor things) but am also getting this error consistently on testnet3 data when testing the binaries manually.

`Deserialize error from connection 127.0.0.1:8338 ud deser h 68088 nttl 23 targets 194 UtxoData[0] err bh 2400000001000213de00000000389758f0001976a9143bb508bf21c15119c134 op 0000000000000000000000000000000000000000ac88affafe952255d81c4525:0 pksize 33706 byte too long`

Can you confirm this?

**Comment by adiabat** [comment]
OK this now works for regtest and testnet3 on my end, would be great if others can try it out.  I'll merge it if there are no other problems found.

File: bridgenode/config.go [hunk 1]
```diff
@@ -18,8 +18,13 @@ The bridgenode server generates proofs and serves to the CSN node.
 OPTIONS:
   -net=mainnet                 configure whether to use mainnet. Optional.
   -net=regtest                 configure whether to use regtest. Optional.
+<<<<<<< HEAD
```
Comment (kcalvinalvin) [inline bridgenode/config.go:21]: Should remove the git merge conflicts here

File: bridgenode/config.go [hunk 2]
```diff
@@ -42,14 +47,12 @@ var (
...
+		`quit generating proofs after the given block height. (meant for testing)`)
 	cowMaxCache = argCmd.Int("cowmaxcache", 4000,
 		`how much memory to use in MB for the copy-on-write forest`)
-	memTTLdb = argCmd.Bool("memttldb", false,
-		`keep a cache of the ttldb in memory.`)
-	allInMemTTLdb = argCmd.Bool("allttldbinmem", false,
-		`keeps the entire ttldb in memory. Uses up a lot of memory.`)
-	quitAtCmd = argCmd.Int("quitat", -1,
-		`quit generating proofs after the given block height. (meant for testing)`)
+	memTTL = argCmd.Bool("memttl", false,
```
Comment (kcalvinalvin) [inline bridgenode/config.go:54]: If we're taking out the option for caching in ram, this flag should also be gone.

File: bridgenode/config.go [hunk 3]
```diff
@@ -208,16 +216,13 @@ type Config struct {
...
 
 	// how much cache to allow for cowforest
 	cowMaxCache int
 
-	// keep the ttldb in memory
-	memTTLdb bool
-
-	// only keep the ttldb in memory
-	allInMemTTLdb bool
+	// keep ttls in memory
```
Comment (kcalvinalvin) [inline bridgenode/config.go:224]: If we're taking out the option for caching in ram, this field should be gone in the config

File: bridgenode/config.go [hunk 4]
```diff
@@ -301,13 +306,7 @@ func Parse(args []string) (*Config, error) {
...
 	cfg.TraceProf = *traceCmd
 	cfg.ProfServer = *profServerCmd
-	cfg.memTTLdb = *memTTLdb
-
-	// If allInMemTTLdb flag was given, the ttldb has to be kept in ram
-	if *allInMemTTLdb {
-		cfg.memTTLdb = true
-	}
-	cfg.allInMemTTLdb = *allInMemTTLdb
+	cfg.memTTL = *memTTL
```
Comment (kcalvinalvin) [inline bridgenode/config.go:309]: Refer to comments before. This should be gone if we're getting rid of caching

File: bridgenode/flatfileworker.go [hunk 5]
```diff
@@ -124,29 +124,42 @@ func flatFileWorker(
...
 			err = ff.writeTTLs(ttlRes)
 			if err != nil {
 				panic(err)
 			}
-		case undo := <-undoChan:
-			err = uf.writeUndoBlock(undo)
-			if err != nil {
-				panic(err)
-			}
+			// case undo := <-undoChan:
```
Comment (kcalvinalvin) [inline bridgenode/flatfileworker.go:158]: I'm assuming commenting this code out is temporary?

File: bridgenode/flatfileworker.go [hunk 6]
```diff
@@ -165,38 +178,41 @@ func (ff *flatFileState) ffInit() error {
...
-		ff.currentOffset, _ = ff.proofFile.Seek(0, 2)
+		ff.currentOffset, err = ff.proofFile.Seek(0, 2)
+		if err != nil {
+			return err
+		}
 
 	} else { // first time startup
 		// there is no block 0 so leave that empty
-		fmt.Printf("setting h=1\n")
+		// fmt.Printf("setting h=1\n")
```
Comment (kcalvinalvin) [inline bridgenode/flatfileworker.go:207]: Should delete this comment

File: bridgenode/flatfileworker.go [hunk 7]
```diff
@@ -244,29 +260,32 @@ func (ff *flatFileState) writeUndoBlock(ub accumulator.UndoBlock) error {
...
-	ff.currentHeight++
+	ff.finishedHeight++
 
 	ff.fileWait.Done()
 
 	return nil
 }
 
 func (ff *flatFileState) writeProofBlock(ud btcacc.UData) error {
+	// fmt.Printf("udata height %d flat file height %d\n",
```
Comment (kcalvinalvin) [inline bridgenode/flatfileworker.go:271]: Should delete these comments

File: bridgenode/flatfileworker.go [hunk 8]
```diff
@@ -278,56 +297,95 @@ func (ff *flatFileState) writeProofBlock(ud btcacc.UData) error {
...
+		if loc == 297548271 {
+			fmt.Printf("loc %d destroyHeight %d c.createHeight %d c.indexWithinBlock %d\n",
+				loc, ttlRes.destroyHeight, c.createHeight, c.indexWithinBlock)
+		}
+
+		_, err := ff.proofFile.ReadAt(readEmpty[:], loc)
 		if err != nil {
 			return err
 		}
+		// fmt.Printf("dest %d at loc %d = (heightOffset[%d] = %d) + 16 + (idx %d *4) \n",
```
Comment (kcalvinalvin) [inline bridgenode/flatfileworker.go:369]: Should delete these comments

File: bridgenode/flatfileworker.go [hunk 9]
```diff
@@ -278,56 +297,95 @@ func (ff *flatFileState) writeProofBlock(ud btcacc.UData) error {
...
+		// ff.heightOffsets[c.createHeight], c.indexWithinBlock)
+
+		if readEmpty != expectedEmpty {
+			return fmt.Errorf("writeTTLs Wanted to overwrite byte %d with %x "+
+				"but %x was already there. desth %d createh %d idxinblk %d",
+				loc, ttlArr, readEmpty, ttlRes.destroyHeight,
+				c.createHeight, c.indexWithinBlock)
+		}
 
+		// fmt.Printf("  writeTTLs overwrite byte %d with %x "+
```
Comment (kcalvinalvin) [inline bridgenode/flatfileworker.go:380]: Should delete these comments

**Review by kcalvinalvin — CHANGES_REQUESTED** [review]

**Comment by kcalvinalvin** [comment]
Ok so the review doesn't really matter as much as commit `f051a5381c22be7a5d66947923fff6be6b51b3f7` doesn't restart for me.

Log:
```
Finished block 48000 of max 1923423                  
^CUser exit signal received. Exiting...
finished reading blocks, last height 48001                 
Done writing. Height 48001 Forest: can't print 50285 leaves. roots:
        0 8a701e97b83879ad08620df1
        1 13247ed533861d880ba53580             
        2 558cf1268abd9f3d5a68b37f
        3 3d5eb15c278c5e8184bf838d    
        4 6752c460effe8486bfa691f6
        5 12a9c1b34c6ab06196e4c3df             
        6 f912e22de9a8d0ccc1f0bb16
        7 5a405d0ce47f9a89aad01772                                                                                      
[N] calvin@bitcoin ~/b/g/u/g/s/g/m/u/c/utreexoserver ((f051a538…))> ./run.sh
Has access to forest, resuming
restore height 48001                                                                                                    
Starting forest: can't print 50285 leaves. roots:                                                                       
        0 8a701e97b83879ad08620df1
        1 13247ed533861d880ba53580             
        2 558cf1268abd9f3d5a68b37f
        3 3d5eb15c278c5e8184bf838d                                                                                      
        4 6752c460effe8486bfa691f6                                                                                      
        5 12a9c1b34c6ab06196e4c3df                                                                                      
        6 f912e22de9a8d0ccc1f0bb16
        7 5a405d0ce47f9a89aad01772                                                                                      
                                                                                                                        
Building Proofs and ttls...                                                                                             
panic: writeTTLs Wanted to overwrite byte 10263779 with 00000003 but 2d3038c3 was already there. desth 48022 createh 48019 idxinblk 20
                                                            
goroutine 15 [running]:          
github.com/mit-dci/utreexo/bridgenode.FlatFileWriter(0xc0004fc240, 0xc0004fc1e0, 0xc00001e378, 0x13, 0xc00001a420, 0x22, 0xc000020800, 0x31, 0xc00001e390, 0x12, ...)
        /home/calvin/bitcoin-projects/go/utreexo/go/src/github.com/mit-dci/utreexo/bridgenode/flatfileworker.go:123 +0x53e
created by github.com/mit-dci/utreexo/bridgenode.BuildProofs     
        /home/calvin/bitcoin-projects/go/utreexo/go/src/github.com/mit-dci/utreexo/bridgenode/genproofs.go:117 +0x3a8
[I] calvin@bitcoin ~/b/g/u/g/s/g/m/u/c/utreexoserver ((f051a538…)) [2]> git log
```

**Comment by adiabat** [comment]
Tested with all of testnet data, looks fine.  There might be other bugs but will merge this now.

**Pull request merged**